### ISSUE TITLE
[Suggestion] HiFiGAN implementation in TTS

### ISSUE THREAD

**Issue by george-roussos** [mozilla/TTS#562]
Recently, a paper on HiFiGAN was published and the authors achieved results better than MelGAN and PWGAN also. Their implementation is here https://github.com/jik876/hifi-gan. I tried it myself and compared to all MelGAN variants and PWGAN, HiFiGAN gave me the best results. Samples can be found here https://jik876.github.io/hifi-gan-demo/. The only part where it does not sound good is when the speaker breathes, which is something I have faced as an issue in all MelGAN implementations. I think it would be a great addition to the TTS repo and with @erogol's knowledge maybe the results could be even better.

Me and @Edresson made some changes to the code to adjust it to Mozilla TTS spectrograms and the fork is here https://github.com/george-roussos/hifi-gan. The code is not very efficient yet, but it works and in case there is interest, it should be good starting ground. Also, if there is interest and a need of more hands, I can look into it myself and send a PR. But I don't know when I would have the time. I also want to train a universal HiFiGAN using LibriTTS.

**Comment by thorstenMueller** [comment]
Thanks @george-roussos.
Samples sound very good (just listened to single speaker ljspeech version). What's the difference between HifiGAN v1, v2 and v3?

@domcross would the be worth a try for our thorsten dataset?

**Comment by erogol** [comment]
It should be quite easy to implement it in here. We already have the barebone training code. Pretty much we need to implement generator and discriminator and allow multiple discriminators to be used in the training code and that would be all. 

Also, there are some things to consider. HiFiGAN is faster than MelGAN but not faster than the multiband version.  I cannot compare MOS values since they use different datasets. However, using the same discriminators as HiFiGAN, one can achieve better results with MultiBand-MelGAN. Basically, they use all the discriminators and the loss functions we have in combination with a different generator architecture.

**Comment by george-roussos** [comment]
> Thanks @george-roussos.
> Samples sound very good (just listened to single speaker ljspeech version). What's the difference between HifiGAN v1, v2 and v3?
> 
> @domcross would the be worth a try for our thorsten dataset?

They aim for shorter synthesis times by reducing the hidden dimensions 😀 I have only used V1. It is longer than MelGAN and shorter than PWGAN.


> It should be quite easy to implement it in here. We already have the barebone training code. Pretty much we need to implement generator and discriminator and allow multiple discriminators to be used in the training code and that would be all.
> 
> Also, there are some things to consider. HiFiGAN is faster than MelGAN but not faster than the multiband version. I cannot compare MOS values since they use different datasets. However, using the same discriminators as HiFiGAN, one can achieve better results with MultiBand-MelGAN. Basically, they use all the discriminators and the loss functions we have in combination with a different generator architecture.

That is so nice to hear! Thanks for the insight. I think the multiple disciminators did it. The one thing HiFiGAN was better at (at least for my dataset) was the lower frequencies when the speaker glottalised. It is the only vocoder so far that was able to model it naturally. I had problems otherwise with every GAN and WaveGrad, even. Also, I agree with you about MultiBand MelGAN. If you think that a combination may work better, it is worth to try. Also, what do you think about PWGAN generator (since it is WaveNet inspired) and the multiple discriminators?

The way that I did training was what suggested on paper. First I trained on raw spectrograms for 1M steps and then finetuned using GTA extracted using my TTS.

**Comment by erogol** [comment]
> The way that I did training was what suggested on paper. First I trained on raw spectrograms for 1M steps and then finetuned using GTA extracted using my TTS.

That also was another difference for HifiGAN. Have you ever trained MelGAN with the same approach?

**Comment by george-roussos** [comment]
> > The way that I did training was what suggested on paper. First I trained on raw spectrograms for 1M steps and then finetuned using GTA extracted using my TTS.
> 
> That also was another difference for HifiGAN. Have you ever trained MelGAN with the same approach?

No sorry, so I don't know what results I could have gotten - however, with HiFiGAN (v1) performance on validation set (before finetuning on TTS spectrograms) was much better than MelGAN val or PWGAN val. It was so good that the only noticeable artifact was in sibilants. So I think HiFiGAN would have beat them.

The only thing that I tried was train PWGAN with TTS spectrograms from scratch. It was horrible.

**Comment by OnceJune** [comment]
> However, using the same discriminators as HiFiGAN, one can achieve better results with MultiBand-MelGAN. Basically, they use all the discriminators and the loss functions we have in combination with a different generator architecture.

Do you mean by keeping MBMelGAN's gen net unchanged, bug use HiFiGAN's dis nets(MPD, MSD), and use mel spec loss instead of stft loss as https://github.com/jik876/hifi-gan/blob/master/train.py?

**Comment by thorstenMueller** [comment]
I'd like to keep this issue open, because of the good compromise between quality and rtf (real time factor). Not sure if the implementation of @george-roussos, @Edresson and @SanjaESC could be easily transfered in tts repo code.

**Comment by faranaziz** [comment]
Hello, 
I am trying the hi-hi gan  from this repo:
https://github.com/george-roussos/hifi-gan

How to generate the files for the text to mel part? 
Thanks you

**Comment by erogol** [comment]
Following #661 I close this one.
